test/test_mail_helpers.py

```diff
@@ -379,6 +379,224 @@ def test_error_is_not_raised_on_to_emails_includes_bcc_cc(self):
+
+    def test_personalization_ccs_setter_filters_out_duplicate_cc_emails_ignoring_case(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        cc_email = Cc('test+cc0@example.com', 'Example Cc Name 0')
+        cc_email_with_caps = Cc('test+CC0@example.com', 'Example Cc Name 0')
+        p.ccs = [cc_email, cc_email_with_caps]
+
+        self.assertEqual([cc_email.get()], p.ccs)
+
+    def test_personalization_bccs_setter_filters_out_duplicate_dict_emails(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        bcc_emails = [{ 'email': 'test+bcc0@example.com', 'name': 'Example Bcc Name 0' }] * 2
+        p.bccs = bcc_emails
+
+        self.assertEqual([bcc_emails[0]], p.bccs)
+
+    def test_personalization_bccs_setter_filters_out_duplicate_dict_emails_ignoring_case(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        bcc_email = { 'email': 'test+bcc0@example.com', 'name': 'Example Bcc Name 0' }
+        bcc_email_with_caps = { 'email': 'test+BCC0@example.com', 'name': 'Example Bcc Name 0' }
+        bcc_emails = [bcc_email, bcc_email_with_caps]
+        p.bccs = bcc_emails
+
+        self.assertEqual([bcc_email], p.bccs)
+
+    def test_personalization_bccs_setter_filters_out_duplicate_bcc_emails(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        bcc_emails = [Bcc('test+bcc0@example.com', 'Example Bcc Name 0')] * 2
+        p.bccs = bcc_emails
+
+        self.assertEqual([bcc_emails[0].get()], p.bccs)
+
+    def test_personalization_bccs_setter_filters_out_duplicate_bcc_emails_ignoring_case(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        bcc_email = Bcc('test+bcc0@example.com', 'Example Bcc Name 0')
+        bcc_email_with_caps = Bcc('test+BCC0@example.com', 'Example Bcc Name 0')
+        p.bccs = [bcc_email, bcc_email_with_caps]
+
+        self.assertEqual([bcc_email.get()], p.bccs)
+
+    def test_personalization_add_to_filters_out_duplicate_to_emails(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        to_email = To('test+to0@example.com', 'Example To Name 0')
+        p.add_to(to_email)
+        p.add_to(to_email)
+
+        expected = [to_email.get()]
+
+        self.assertEqual(expected, p.tos)
+
+    def test_personalization_add_bcc_filters_out_duplicate_bcc_emails(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        bcc_email = Bcc('test+to0@example.com', 'Example To Name 0')
+        p.add_bcc(bcc_email)
+        p.add_bcc(bcc_email)
+
+        expected = [bcc_email.get()]
+
+        self.assertEqual(expected, p.bccs)
+
+    def test_personalization_add_cc_filters_out_duplicate_cc_emails(self):
+        self.maxDiff = None
+
+        p = Personalization()
+        cc_email = Cc('test+to0@example.com', 'Example To Name 0')
+        p.add_cc(cc_email)
+        p.add_cc(cc_email)
+
+        expected = [cc_email.get()]
+
+        self.assertEqual(expected, p.ccs)
```

Comment:
These look to be duplicates of:

- https://github.com/sendgrid/sendgrid-python/pull/924/files#diff-dfb342158840d4c80c26b95c9be2c3fa7bba624850d288105c5462ce991014daR382
- https://github.com/sendgrid/sendgrid-python/pull/924/files#diff-dfb342158840d4c80c26b95c9be2c3fa7bba624850d288105c5462ce991014daR403
- https://github.com/sendgrid/sendgrid-python/pull/924/files#diff-dfb342158840d4c80c26b95c9be2c3fa7bba624850d288105c5462ce991014daR424

Reply:
Semantically yes, but they are testing a different helper method.  Would you prefer to have them to make sure the helper methods that `add_email` uses works properly (`add_to`, `add_bcc`, and `add_cc`), or just have unit tests for `add_email` and not `add_to` and etc.?

Reply:
I agree we should test them separately to ensure all of them work as expected, since it's all unique code.